### PR TITLE
Add code commentary to LibExecution

### PR DESCRIPTION
**Pull request by StachuDotNet**
## What is the problem/goal being addressed?
Increase in-code documentation of LibExecution, for future developers' understanding.
Also includes some no-op changes to other projects.

## What is the solution to this problem?
- adjust existing commentary from //-style to ///-style, so IDEs show the the descriptions embedded in their UIs
- add additional commentary (and some questions) where relevant

### CODE REVIEW

File: fsharp-backend/src/LibExecution/ProgramTypes.fs [hunk 1]
```diff
@@ -343,7 +341,7 @@ and Pattern =
 
 
 
-
+// TODO: this seems to be redundant with separate Shortcuts.fs file
```
Comment (StachuDotNet) [inline fsharp-backend/src/LibExecution/ProgramTypes.fs:344]: @pbiggar can/should we delete one of these Shortcuts modules?

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 2]
```diff
@@ -169,25 +170,38 @@ module DDateTime =
...
 
-// This Expr is the AST, expressing what the user sees in their editor.
+/// The AST, expressing what the user sees in their editor.
 type Expr =
   | EInteger of id * int64
   | EBool of id * bool
   | EString of id * string
   | ECharacter of id * string
-  | EFloat of id * double // first string might have a sign in it
+  | EFloat of id * double // first string might have a sign in it <- what?
```
Comment (StachuDotNet) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:179]: What does this mean?

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 3]
```diff
@@ -169,25 +170,38 @@ module DDateTime =
...
   | ENull of id
   | EBlank of id
+
+  /// Composed of binding name, expression, and ???
```
Comment (StachuDotNet) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:183]: what's the 2 expressions here? I assume one of them is _the_ expression being bound, but not sure what the other is.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 4]
```diff
@@ -169,25 +170,38 @@ module DDateTime =
...
   | ELet of id * string * Expr * Expr
+
+  /// Composed of condition, expr if true, and expr if false
   | EIf of id * Expr * Expr * Expr
+
+  /// Composed of a parameters * the expression itself
   | ELambda of id * List<id * string> * Expr
+
+  /// ?
```
Comment (StachuDotNet) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:192]: What is EFieldAccess?

File: fsharp-backend/src/LibExecution/Shortcuts.fs [hunk 5]
```diff
@@ -63,7 +68,7 @@ let rec toStringRepr (e : Expr) : string =
   | EList (_, exprs) ->
     let exprs = List.map r exprs |> String.concat "; "
     $"eList [{exprs}]"
-  | _ -> $"Bored now: {e}"
+  | _ -> $"Bored now: {e}" // should we finish this? TODO check usages
```
Comment (StachuDotNet) [inline fsharp-backend/src/LibExecution/Shortcuts.fs:71]: is anything impacted by this function being incomplete?

File: fsharp-backend/src/HttpMiddleware/Init.fs [hunk 6]
```diff
@@ -1,9 +1,8 @@
 module HttpMiddleware.Init
 
-// Initialize HttpMiddleware
-
 open Prelude
 
+/// Initialize HttpMiddleware
```
Comment (pbiggar) [inline fsharp-backend/src/HttpMiddleware/Init.fs:5]: This was supposed to be the heading for the file.

File: fsharp-backend/src/HttpMiddleware/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# HttpMiddleware
+
+HTTP Middleware modules to support BwdServer.
+
+Designed to move as much of the HTTP framework into "user space"
+(or something which could potentially be user space in the future)
+as possible.
+
+This library only exists so that we can separate LibBackendStdLib from LibBackend
+does that^ comment (pull from fsproj) still make sense?
```
Comment (pbiggar) [inline fsharp-backend/src/HttpMiddleware/README.md:10]: Maybe. I don't really remeber why it's true though.

File: fsharp-backend/src/LibExecution/ProgramTypes.fs (hunk 1 above)
Comment (pbiggar) [reply fsharp-backend/src/LibExecution/ProgramTypes.fs:344]: Shortcuts.fs is for RuntimeTypes, this is for ProgramTypes. I agree a refactor is in order! We may be able to get rid of ProgramTypes shortcuts as we have the "parser" now.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs (hunk 2 above)
Comment (pbiggar) [reply fsharp-backend/src/LibExecution/RuntimeTypes.fs:179]: Nothing, it refers to an old way I was trying to represent this.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs (hunk 3 above)
Comment (pbiggar) [reply fsharp-backend/src/LibExecution/RuntimeTypes.fs:183]:
```
let str = expr1
expr2
```

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs (hunk 4 above)
Comment (pbiggar) [reply fsharp-backend/src/LibExecution/RuntimeTypes.fs:192]:
```
someExpr.fieldName
```

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 8]
```diff
@@ -364,9 +391,9 @@ and DType =
 
 
 
-// Record the source of an incomplete or error. Would be useful to add more
-// information later, such as the iteration count that let to this, or
-// something like a stack trace
+/// Record the source of an incomplete or error. Would be useful to add more
+/// information later, such as the iteration count that let to this, or
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:395]:
```suggestion
/// information later, such as the iteration count that led to this, or
```

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 9]
```diff
@@ -785,12 +844,18 @@ type BuiltInFn =
     previewable : Previewable
     deprecated : Deprecation
     sqlSpec : SqlSpec
+
+    // TODO: reevaluate usefulness of below comment. at least rephrase.
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:848]: I think this is still accurate, though the implementations are now in BackendOnlyStdlib, not LibBackend.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 10]
```diff
@@ -865,28 +934,44 @@ and ExceptionReporter = ExecutionState -> Metadata -> exn -> unit
...
+    /// but with recursion that causes infinite recursion.
+    /// </remarks>
     callstack : Set<FQFnName.T>
-    // Whether the currently executing code is really being executed (as
-    // opposed to being executed for traces)
+
+    /// Whether the currently executing code is really being executed
+    /// (as opposed to being executed for traces)
+    /// TODO: find a better word than 'executed' above. Generally clarify meaning.
     onExecutionPath : bool }
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:975]: "whether the current code is being previewed"? Could rename the var too.

File: fsharp-backend/src/LibExecution/Shortcuts.fs (hunk 5 above)
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/Shortcuts.fs:71]: No, this was purely for writing code from existing expressions, which I wanted to do for middleware. I think we'd use FSharpToExpr for this now. I haven't used this in forever, no objection to deleting it.

File: fsharp-backend/src/LibExecution/VendoredTablecloth.fs [hunk 11]
```diff
@@ -1,11 +1,15 @@
+/// A forked subset of Tablecloth
+///
+/// Kept separate from Tablecloth as, in LibExecution, we want to be very careful
+/// about backward compatibility. Tablecloth is still very much in flux, so
+/// we want to avoid using it in case something happens that breaks lots of things.
+///
+/// Our rule for is to not use Tablecloth in LibExecution.
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/VendoredTablecloth.fs:7]:
```suggestion
/// Our rule is to not use Tablecloth in LibExecution.
```

**Review by pbiggar — CHANGES_REQUESTED** [review]
Looking good.

File: fsharp-backend/src/LibExecution/ProgramTypes.fs (hunk 1 above)
Comment (StachuDotNet) [reply fsharp-backend/src/LibExecution/ProgramTypes.fs:344]: I tried deleting and going down this route, but wasn't confident how to replace the usage within `TestUtils/FSharpToExpr`, so created a separate Issue to consider the path forward here.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 12]
```diff
@@ -845,10 +855,8 @@ type BuiltInFn =
     deprecated : Deprecation
     sqlSpec : SqlSpec
 
-    // TODO: reevaluate usefulness of below comment. at least rephrase.
-    // Functions can be run in JS if they have an implementation in this
-    // LibExecution. Functions whose implementation is in LibBackend can only be
-    // implemented on the server.
+    // Functions can be run in JS if they have an implementation in BackendOnlyStdlib.
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:858]: // Functions can be run in JS if they have an implementation in **LibExecution**.
    // Functions whose implementation is in **BackendOnlyStdLib** can only be implemented on the server.

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs [hunk 13]
```diff
@@ -864,10 +872,9 @@ and Fn =
     previewable : Previewable
     deprecated : Deprecation
     sqlSpec : SqlSpec
-    // TODO: reevaluate usefulness of below comment. at least rephrase.
-    // Functions can be run in JS if they have an implementation in this
-    // LibExecution. Functions whose implementation is in LibBackend can only be
-    // implemented on the server.
+
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/RuntimeTypes.fs:875]: This is duplicated.

**Review by pbiggar — CHANGES_REQUESTED** [review]
Couple of tiny changes left

File: fsharp-backend/src/LibExecution/RuntimeTypes.fs (hunk 13 above)
Comment (StachuDotNet) [reply fsharp-backend/src/LibExecution/RuntimeTypes.fs:875]: It was duplicated in the original code, so I kept it such. I can remove one, though - which?

Comment (pbiggar) [reply fsharp-backend/src/LibExecution/RuntimeTypes.fs:875]: Either is fine.

**Comment by StachuDotNet** [comment]
https://app.circleci.com/pipelines/github/darklang/dark/2330/workflows/e9c5ba49-f24b-4bb1-bcc5-894f0d00770c/jobs/24412
some unexpected test failures. I'm guessing re-running the build would magically resolve these, but I don't have permissions to hit 'rerun' (beyond pushing a bogus commit)

**Comment by pbiggar** [comment]
This is good to merge, right?

**Comment by StachuDotNet** [comment]
Yes

**Pull request merged**